### PR TITLE
feat(web-analytics): Add custom channel types

### PR DESCRIPTION
**Pull request by robbie-c**
## Problem

https://github.com/PostHog/posthog/issues/21195

This PR lets people add custom channel types. They take the form of rules, where a list of rules might be (in plain english)

* If the utm_source is 'ai' then the channel type is 'AI'
* If the utm_medium is 'newsletter' or the referring domain is 'substack.com' then the channel type is 'Newsletter'

These rules are a new field HogQLQueryModifiers, and are added to the HogQL for session.$channel_type when it is queried. The first matching rule wins, and if no rules match then we use the defaul channel type logic.

## Changes

Add the query-side changes, and placeholder front end UI to let people customize channel types.

It's behind a FF, and a follow up PR will make the actual UI ;)

Here's the temporary UI, but please don't judge me :)
Screenshot 2024-11-13 at 20 30 49

And here's that appearing in a query
Screenshot 2024-11-13 at 20 31 34

Custom channel types in the channel type filter UI:
Screenshot 2024-11-13 at 20 51 56

## Does this work well for both Cloud and self-hosted?

Yes

## How did you test this code?

Added a couple of tests in the channel type code

### CODE REVIEW

**Review by Twixes — COMMENTED** [review]
Looked through the frontend code, a couple of comments. Got zero context on the feature, so didn't look at the backend logic

File: frontend/src/scenes/settings/environment/ChannelType.tsx [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    }
+
+    return (
+        <>
+            <p>Set your custom channel type</p>
+            <LemonInput
+                value={customChannelTypeRules}
+                onChange={setCustomChannelTypeRules}
+                placeholder="Enter JSON array of custom channel type rules"
+            />
```
[the remark this answers is not in the document]
Comment (robbie-c) [reply frontend/src/scenes/settings/environment/ChannelType.tsx:40]: Ah fair - this is really just a placeholder that won't be seen by anyone, it only exists to split things into 2 PRs rather than one big one, so I'm not too fussed about it being suboptimal.

File: frontend/src/queries/schema.ts [hunk 2]
```diff
@@ -2283,3 +2284,38 @@ export enum AssistantGenerationStatusType {
...
+    NotRegex = 'not_regex',
+}
+
+export interface CustomChannelCondition {
+    key: CustomChannelField
+    value?: string | string[]
+    op: CustomChannelOperator
+}
+
+export enum CustomChannelCombiner {
```
[the remark this answers is not in the document]
Comment (robbie-c) [reply frontend/src/queries/schema.ts]: Makes sense, done

File: frontend/src/scenes/settings/environment/ChannelType.tsx [hunk 3]
```diff
@@ -0,0 +1,48 @@
+import { useActions, useValues } from 'kea'
+import { LemonButton } from 'lib/lemon-ui/LemonButton'
+import { eventUsageLogic } from 'lib/utils/eventUsageLogic'
+import { useState } from 'react'
+import { teamLogic } from 'scenes/teamLogic'
+
+import { CustomChannelRule } from '~/queries/schema'
+import { LemonInput } from 'lib/lemon-ui/LemonInput'
+
+export function ChannelType(): JSX.Element {
```
Comment (robbie-c) [inline frontend/src/scenes/settings/environment/ChannelType.tsx:10]: This code is placeholder and behind a FF, so it's ok that's imperfect. It only exists to split this PR into 2 rather 1 big PR

File: posthog/hogql/database/schema/channel_type.py [hunk 4]
```diff
@@ -38,29 +56,130 @@ def create_initial_domain_type(name: str):
...
+                gad_source=ast.Call(name="toString", args=[ast.Field(chain=["properties", "$initial_gad_source"])]),
             ),
-            gclid=ast.Call(name="toString", args=[ast.Field(chain=["properties", "$initial_gclid"])]),
-            gad_source=ast.Call(name="toString", args=[ast.Field(chain=["properties", "$initial_gad_source"])]),
+            custom_rules=custom_rules,
         ),
     )
 
 
+def custom_condition_to_expr(
```
Comment (robbie-c) [inline posthog/hogql/database/schema/channel_type.py:78]: This is very similar to the code in hogql/property.py, but had enough minor differences that I thought it would be better to WET rather than trying to shoehorn in those differences.

Comment (pauldambra) [reply posthog/hogql/database/schema/channel_type.py:78]: "a little duplication is better than the wrong abstraction" 👍

File: frontend/src/lib/utils/eventUsageLogic.ts [hunk 5]
```diff
@@ -810,6 +811,9 @@ export const eventUsageLogic = kea<eventUsageLogicType>([
         reportSessionTableVersionUpdated: async ({ version }) => {
             posthog.capture('session table version updated', { version })
         },
+        reportCustomChannelTypeRulesUpdated: async ({ numRules }) => {
+            posthog.capture('custom channel type rules updated', { numRules })
+        },
```
Comment (pauldambra) [inline frontend/src/lib/utils/eventUsageLogic.ts:816]: i sort of hate the eventUsageLogic so would just do this capture in line but 🤷

File: frontend/src/queries/schema.ts [hunk 6]
```diff
@@ -2521,3 +2522,33 @@ export enum AssistantGenerationStatusType {
...
+export enum CustomChannelOperator {
+    Exact = 'exact',
+    IsNot = 'is_not',
+    IsSet = 'is_set',
+    IsNotSet = 'is_not_set',
+    IContains = 'icontains',
+    NotIContains = 'not_icontains',
+    Regex = 'regex',
+    NotRegex = 'not_regex',
+}
```
Comment (pauldambra) [inline frontend/src/queries/schema.ts:2542]: is this just the same as property operators? does it need to be its own declaration? or could be a `Pick<>` or `Omit<>`

File: posthog/hogql/database/schema/channel_type.py [hunk 7]
```diff
@@ -38,29 +56,130 @@ def create_initial_domain_type(name: str):
...
+                    ast.Call(name="or", args=[custom_condition_to_expr(expr, val, condition.op) for val in value])
+                )
+        else:
+            conditions.append(custom_condition_to_expr(expr, value, condition.op))
+    if len(conditions) == 0:
+        return ast.Constant(value=True)
+    elif len(conditions) == 1:
+        return conditions[0]
+    else:
+        return ast.Call(name=custom_rule.combiner.lower(), args=conditions)
```
Comment (pauldambra) [inline posthog/hogql/database/schema/channel_type.py:178]: this is a really cool feature because it's super close to adding the specification pattern to our filters for arbitrary custom things not just channel types

that'd be scope creep here but so i'd be tempted to call this "specification" not "combiner" but that's not-blocking now

**Review by pauldambra — APPROVED** [review]
this is really cool and behind a flag so 🚢
